[Mondrian] Implement viewer for Mondrian files

ai-moiseev opened the pull request:
This commit implements custom editor which provides graphical view of Mondrian JSON allocation traces (`.tracealloc.json`).

Issue: #395 
Draft: #452 

ONE-vscode-DCO-1.0-Signed-off-by: Alexander Moiseev <a.moiseev@samsung.com>

## Notes

Currently implemented features: 
* JSON input data parsing
* round-robin colorized memory blocks output
* ability to scale horizontally/vertically
* switching between memory segments.

More features such as model trim, mouse hover info, rulers, more coloring modes etc. will be added in new PRs, #395 description contains feature roadmap.

Example data: 
[example.tracealloc.json.txt](https://github.com/Samsung/ONE-vscode/files/8531449/example.tracealloc.json.txt) (remove .txt extension for plugin to properly detect .json)

## Screenshots

Scaled up memory zone view:

![image](https://user-images.githubusercontent.com/67966333/164460590-4a6e2835-f0cd-4ffe-8bd4-c545233c69cd.png)

Model "bird's eye" overview:

![image](https://user-images.githubusercontent.com/67966333/164460694-3142a36a-aaf7-4dff-8acd-24af649bde10.png)

Light theme support:

![image](https://user-images.githubusercontent.com/67966333/164460878-a77e31e3-cb16-43e1-85b7-fd51d5bf5215.png)

llFreetimell commented on media/Mondrian/mondrianViewer.js:
```diff
@@ -0,0 +1,265 @@
+
+  const viewportCycles = viewer.viewportMaxCycle - viewer.viewportMinCycle;
+  for (const [i, alloc] of data.segments[viewer.activeSegment].allocations.entries()) {
+    if (alloc.alive_from > viewer.viewportMaxCycle) {
```
As I am not good at javascript,  please understand me even I ask some stupid questions :)
( I am glad that you created a good example with custom editor because I am working with custom editor too :D )

---

From line 131, following codes can be found.

```js
  for (alloc of data.segments[viewer.activeSegment].allocations) {
    if (alloc.alive_till > totalCycles) {
      totalCycles = alloc.alive_till;
    }
    // ...
viewer.viewportMaxCycle = totalCycles;
```

Then... is this case can be occurred?

ai-moiseev replied:
Oh... this check was added in advance before implementation of the next feature - the ability to show only a part of the model (large models with 100k+ cycles can slow things down quite a bit... that's the purpose of `viewportMinCycle` / `viewportMaxCycle` properties - this check skips allocations outside of selected range). I haven't implemented it yet (I need to implement interactive range selector first), so I plan to add this feature in the next PR :)

llFreetimell commented on src/Mondrian/MondrianEditor.ts:
```diff
@@ -0,0 +1,105 @@
+      </head>
+      <body>
+        <div class="mondrian-layout">
+          <div class="mondrian-scrollbar"></div>
```
I cannot find any information about this div from draft PR :(
I think this is some kind of scrollbar with preview but not sure... is it NYI?

ai-moiseev replied:
Yes, this is a placeholder :(
I've added NYI note :)

llFreetimell commented on src/Mondrian/MondrianEditor.ts:
```diff
@@ -0,0 +1,105 @@
+            <div class="mondrian-info">
+              <b>Memory:</b> <span class="mondrian-info-memory-size">0</span> |
+              <b>Cycles:</b> <span class="mondrian-info-cycle-count">0</span> |
+              <b>Segment:</b> <select class="mondrian-segment-picker"></select>
```
Just FYI, there is `vscode-dropdown` for `select` in VSCode :D

ai-moiseev replied:
`vscode-dropdown` looks nice! I've added vscode toolkit dependency and replaced `<select>` with `<vscode-dropdown>` :)

hyunsik-yoon commented:
Could you also add json file into, e.g., `res/sample`? Thanks.

hyunsik-yoon commented:
@ai-moiseev Thanks for your work!
I am trying to understand the whole flow. (based on https://github.com/Samsung/ONE/issues/8743)

1. vscode runs `onecc` to request memory measurement for backend Z.
2. `onecc` calls some tool defined in backend Z.
3. some tool defined in backend Z produces memory measurement result in format of X.
4. `onecc` produces the result in json format in this PR.
5. vscode renders the json.

Some part of step 1 (`onecc` param), step 2, 3, and 4 are not a scope of _ONE-vscode_ but a scope of _ONE_. 
And, step 5 (this PR, in scope of ONE-vscode) is meaningful only if step 1, 2, 3, and 4 are defined and implemented.
If there's discussion in ONE repo regarding those steps, could you let us know?

ai-moiseev commented:
@llFreetimell As you know, there was a discussion here: https://github.com/Samsung/ONE/issues/8723
It was stated that support for memory allocation format handling is outside of ONE's scope :( So I've decided to provide implementation in backend and ONE-vscode and to continue discussion on this topic.

My ONE's proposal initially included JSON file processing and conversion into SVG format, but with introduction of custom editor/viewer in ONE-vscode, we only need to add `onecc` compile option. I'll prepare PR for it and try to get it merged :-)

hyunsik-yoon commented:
@ai-moiseev 
I think there was some gap of approach between our opinions.

What I have been thought was https://github.com/Samsung/ONE-vscode/pull/483#issuecomment-1109210443.

What you've though was (I guess)
- support for memory allocation format handling is outside of ONE's scope

I am OK with your way. I guess that we can have short-term and long-term approach:
- short term: introduce mondrian map of backend in ONE-vscode and backend (not ONE)
- long term: _check_ introduction of mondrian to ONE (ONE is a bridge between ONE-vscode and backends)
  - to align with https://github.com/Samsung/ONE/issues/8743

In this view, I am OK with this PR. :-) Let me know your opinion.

hyunsik-yoon commented:
I would be helpful to have a explanation of schema of json as, e.g., README.md file. Could you write one in, e.g., `src/mondrian`?

ai-moiseev commented:
> I would be helpful to have a explanation of schema of json as, e.g., README.md file. Could you write one in, e.g., `src/mondrian`?

I've added data format specification in [json-schema](https://json-schema.org/) format.

hyunsik-yoon commented:
Oh. Maybe my comment was not complete enough. 😹 
What I am curious is the meaning of each key and value.

E.g., If I want to write json in an editor, 
I have to understand the meaning of "segments", or "size", "allocations", "offset" and so on.
Since this repo is public, anyone who access this repo also needs to understand the meaning of those keys and values.
So could you write down the meaning in README.md ?

If this explanation requires some backend knowledge, you may want to discuss this in [other repo](https://github.sec.samsung.net/RS7-RuntimeNTools/ONE-vscode-TRIV/issues).

```
{
  "schema_version": 1,
  "segments": [
    {
      "name": "Segment1",
      "size": 32768,
      "allocations": [
        { "offset": 0, "size": 8192, "alive_from": 0, "alive_till": 8, "origin": "Node1" },
....
```

ai-moiseev commented:
@hyunsik-yoon I've updated `README.md` and added text description of all fields, PTAL :) This format is intended to be as backend agnostic as possible (although it's usage limited to backends with memory scheduling; not all backends have precise cycle or timing control, so there's abstract 'lifetime' attribute, which can be relative, which is useful for `luci-micro` backend).

hyunsik-yoon commented:
@ai-moiseev Thanks for the README.md.

I tried to write down some suggestion or question of my UI experience. 
Some could be done in this PR or in next PR, depending on your opinion. :-)

- Some people may not know what this json file is. So , when they click the json file, how about showing some title like `Memory map over time` and some description on top of graph?
- After opening json file, right-bottom area shows `memory`, `cycle`, and `segment`. I guess `memory` here means total memory required through whole times? If so, how about using `Total memory` label?
  -  When I select `segment 2`, `memory` is 16384 and I am not sure why. From the graph, it has one segment of 8K size. So would 8K be right for `memory`? (So does segment 1)
- I guess X axis would be time (cycle??) and Y axis would be offset but I am not sure. How about showing the meaning of X and Y axis?
- When I maximize the size of vscode to full screen and click the json file, initial graph is shown in small top left area. Maybe when user clicks json, how about fitting the graph to screen size?
- Some button like 'Fit to screen size' would provide good usability.
- In the json file, segment 1 has 24 cycles and segment 2 has 32 cycles. When I enlarge segment 1 to fit all cycle in width of screen and then select segment 2, segment 2 also shows 24 cycles in screen. So I fit 32 cycles of segment 2 to the width of screen. Then I select segment 1 again and found that graph is now shrunken and I had to enlarge again. I hope zoom level of each segment should be kept separately.
- zooming granularity of '+' and '-' seems kinda large, I guess.

ai-moiseev commented:
@hyunsik-yoon Thank you for a feedback! Yes, explanation UI elements are definitely needed (I planned to add rulers on top and the right and label them as 'time' and 'memory'.

Fit to screen buttons are planned (for X and Y axes) as well as ability to remember zoom and position for each memory segment (I wanted to implement submodel view functionality first and make sure these two features will not conflict with each other).

Zooming step can easily be adjusted :) (now it's 1.0, but could be anything). I'll send update to this PR which will speed it down 2x shortly.

Thank you again for your notes! I would like to quote you in parent issue to keep track of things to implement.

hyunsik-yoon commented on res/samples/traces/sample.tracealloc.json:
```diff
@@ -0,0 +1,22 @@
+    },
+    {
+      "name": "Segment2",
+      "size": 16384,
```
Shouldn't this be 8192?
Could you also check whether total memory at Line 6 is correct?

ai-moiseev replied:
I've reduced segment sizes to precisely fit all the blocks. Anyway, the purpose of those samples is to demonstrate syntax rather than real data, I guess, so the values may be a little weird :)

ai-moiseev commented:
> Some part of step 1 (`onecc` param), step 2, 3, and 4 are not a scope of _ONE-vscode_ but a scope of _ONE_.
> And, step 5 (this PR, in scope of ONE-vscode) is meaningful only if step 1, 2, 3, and 4 are defined and implemented.
> If there's discussion in ONE repo regarding those steps, could you let us know?

@hyunsik-yoon I've created PR which adds `save-allocations` option to `onecc codegen` which should provide a bridge between `ONE-vscode`, `ONE` and backends with allocation trace export support.

llFreetimell reviewed (approved):
LGTM
